### PR TITLE
Move labelpattern

### PR DESCRIPTION
**Pull request by lenhard**
Minor refactoring and moving of classes from logic into model. The `XXXLabelPattern` classes really describe how we serialize and store label patterns (part of the structure of meta data). Hence, they belong into model and have already been almost independent of other classes. The util class with helpful functions regarding label patterns can stay in util.

### CODE REVIEW

File: src/main/java/net/sf/jabref/model/labelpattern/GlobalLabelPattern.java [hunk 1]
```diff
@@ -13,15 +13,31 @@
...
 public class GlobalLabelPattern extends AbstractLabelPattern {
 
+    private static List<String> defaultLabelPattern;
+
+    static {
+        updateDefaultPattern();
+    }
+
+    public static void updateDefaultPattern() {
+        defaultLabelPattern = AbstractLabelPattern.split(JabRefPreferences.getInstance().get(JabRefPreferences.DEFAULT_LABEL_PATTERN));
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/model/labelpattern/GlobalLabelPattern.java]: See #1607. :-) Not sure what is the best merging order...

Comment (lenhard) [inline src/main/java/net/sf/jabref/model/labelpattern/GlobalLabelPattern.java]: @simonharrer Good idea
@oscargus Oh no, now I have already rewritten `GlobalLabelPattern` :-(

However, it seems you did not really touch the class in #1607? From what I have seen and tested with the GUI, you can completely drop the static part of `GlobalLabelPattern`. It is just initialized in two places (the gui and the preferences) and both of these places have access to the gui, so you can provide the correct parameters right away.

Regarding merging order: It prefer mine to be merged first :P Anyway, let Simon decide!

**Comment by simonharrer** [comment]
LGTM

@oscargus how does this affect #1607 ?

**Comment by oscargus** [comment]
No idea. Probably a bit of resolving independent of who merges first. The feeling is that it is easier to merge the PR without moving of files first, but I'm not sure if that is true or if I just want less work. ;-)

**Comment by oscargus** [comment]
By the way, there is a preference setting "Enforce legal characters in BibTeX keys". Is that really needed?

The difference is the characters in `"{}(),\\\"#~^'"` but not in `"{}(),\\\""` so `#~^'`. I'm wondering if we upset many people by always enforcing legal characters?

**Comment by lenhard** [comment]
@oscargus: My changes were quite small and would be relatively easy to redo, so it's ok for me if you go first.

Regarding the BibTex keys: We have already upset at least one person in #1272. We talked about this issue on the devcall right now and arrived at the following decision:
 - The preference setting "Enforce legal characters in BibTeX keys" can be removed, hence we will only enforce `"{}(),\\\"#~^'"`.
 - Key validation should be based on the mode (BibTeX, BibLaTeX). Especially BibLaTeX is more strict and there should be an additional validation that checks for the symbols that are not allowed in BibLaTeX

**Comment by oscargus** [comment]
OK! If anyone could have a look at the PR I'll merge it.

So basically, the (in the PR introduced) boolean argument to `checkLegalKey` should be replaced with isBibLatex? Is it clear what the set of symbols are for the two cases?

**Comment by tobiasdiez** [comment]
And we considered moving the check to the integrity dialog.

**Comment by lenhard** [comment]
Should be good to go. 

Just as a side note: What are the "default" buttons in Preferences-> BibTeX key generator supposed to do? For me, only the button for the default pattern works and actually resets something. The other default buttons basically just delete what I have entered into the corresponding text field. This behavior is independent of this PR.

**Comment by oscargus** [comment]
I think the default is that every type has the same key pattern, so it makes sense that the pattern is cleared when clicking Default.

**Comment by oscargus** [comment]
LGTM! :+1:

**Comment by matthiasgeiger** [comment]
I'm not sure whether we discussed it in this context at the last devcall, but moving those classes will lead to another node in the registry for storing those preferences - if the storing mechanism is not adjusted.

**Comment by lenhard** [comment]
@matthiasgeiger: Are you sure? Didn't that already happen through #1635 by introducing the new `LabelPatternPreferences`? As far as I understand, the old label preferences get lost through this change, but will not so easily get lost in the future, since they are now normal preferences.

**Comment by matthiasgeiger** [comment]
Just tested it: v3.5 and current master reads and writes from/to  `HKEY_CURRENT_USER\SOFTWARE\JavaSoft\Prefs\net\sf\jabref\logic\labelpattern` - your branch reads and writes from/to `HKEY_CURRENT_USER\SOFTWARE\JavaSoft\Prefs\net\sf\jabref\model\labelpattern`

**Comment by matthiasgeiger** [comment]
Caused by: https://github.com/JabRef/jabref/blob/master/src/main/java/net/sf/jabref/preferences/JabRefPreferences.java#L1151

Maybe we should discuss again how to implement this better and provide migrations this time ;-)

**Comment by lenhard** [comment]
Migrations are overrated :-)

We should at least change that to 

`Preferences pre = Preferences.userNodeForPackage(JabRefMain.class);`

to avoid future issues.

Although not part of this PR, let us keep this open and discuss at the next devcall on how to proceed.

**Comment by matthiasgeiger** [comment]
Another topic for the devcall: Can we rename this "LabelPattern..." stuff to "BibtexKeyPattern..." ? - I never understood why this is called "label pattern" :wink:

**Comment by lenhard** [comment]
TODO in new PR after this one is merged:
 - Integrate label pattern preferences tree into global preferences tree
 - Rename label pattern to bibtex key pattern.
 - Possible migration of old key patterns

**Pull request merged**